### DISCUSSION TITLE
Option to skip inlining or reading any env vars at build time

### DISCUSSION THREAD

**Discussion by codan84** [vercel/next.js#66497] [Ideas]
### Goals

1. In proper CI/CD world I should be able to build a Docker image once and promote it through environments
2. This can't be done right now as `next build` tries to read bunch of environment variables (yes I read posts and I don't care about dynamic routes, SSR etc, this should not be my concern) - bottom line is: 1 `next build`, 1 docker image, __ALL__ env vars loaded via  .env at runtime!
4. There is no excuse to not have that option.

### Non-Goals

1. Removing any kind of optimisations
2.
3.

### Background

Right now because I need a bunch of env vars when running `next build`, some of which get inlined in generated code, I am forced to do either one of these:
1. Build per environment (WRONG approach - one should build __once__ and promote given build through environments)
2. Do any of numerous workarounds to avoid env vars being inlined etc (again, WRONG, I should not have to have a cognitive overload of solving this, I should be focusing on resolving business problem)

### Proposal

Give me an option to run build that does not require __ANY__ env vars to be set, no matter how I implement my page.

**Comment by m-sanders** [comment]
This is possible!

You can split the build into two steps by using the next build experimental option documented here: 

https://github.com/vercel/next.js/discussions/46544

The app will still compile build time env var into source so don’t use them and instead use a runtime env option. For App Router you can use the next-runtime-env npm package. Page Router has first party support for runtime env by using the runtimeConfig next.config option.
